File: Advanced/PageAndRDL/OracleDataProvider/C#/OracleDataProvider/DataExtensions.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace GrapeCity.ActiveReports.Samples.OracleDataProvider
{
	/// <summary>
	/// Provides various extension methods for data interfaces.
	/// </summary>
	internal static class DataExtensions
	{
		/// <summary>
		/// Reads all records in specified data reader.
		/// </summary>
		/// <param name="reader">The data reader to iterate records for.</param>
		/// <returns></returns>
		public static IEnumerable<System.Data.IDataRecord> ReadAll(this System.Data.IDataReader reader)
		{
			while (reader.Read())
			{
				yield return reader;
			}
		}

		/// <summary>
		/// Gets field value as string in specified record.
		/// </summary>
		/// <param name="record">The record to get value field in.</param>
		/// <param name="name">The field name to get value for.</param>
		/// <returns>The field value.</returns>
		public static string GetString(this System.Data.IDataRecord record, string name)
		{
			var value = record.GetValue(record.GetOrdinal(name));
			return Convert.ToString(value, CultureInfo.InvariantCulture);
		}

		/// <summary>
		/// Gets field value as integer in specified record.
		/// </summary>
		/// <param name="record">The record to get value field in.</param>
		/// <param name="name">The field name to get value for.</param>
		/// <returns>The field value.</returns>
		public static int GetInt32(this System.Data.IDataRecord record, string name)
		{
			var value = record.GetValue(record.GetOrdinal(name));
			return Convert.ToInt32(value, CultureInfo.InvariantCulture);
		}

		/// <summary>
		/// Gets field value as boolean in specified record.
		/// </summary>
		/// <param name="record">The record to get value field in.</param>
		/// <param name="name">The field name to get value for.</param>
		/// <returns>The field value.</returns>
		public static bool GetBoolean(this System.Data.IDataRecord record, string name)
		{
			var value = record.GetValue(record.GetOrdinal(name));
			return Convert.ToBoolean(value, CultureInfo.InvariantCulture);
		}

		/// <summary>
		/// Return whether the specified field is set to null.
		/// </summary>
		/// <param name="record">The record to get value field in</param>
		/// <param name="name">The field name to get value for.</param>
		/// <returns>true if the specified field is set to null; otherwise, false.</returns>
		public static bool IsDBNull(this System.Data.IDataRecord record, string name)
		{
			var value = record.GetValue(record.GetOrdinal(name));
			return value == DBNull.Value;
		}
	}
}

File: Advanced/PageAndRDL/OracleDataProvider/VB.NET/OracleDataProvider/DataExtensions.vb
Imports System
Imports System.Collections.Generic
Imports System.Globalization
Imports System.Runtime.CompilerServices

Namespace GrapeCity.ActiveReports.Samples.OracleDataProvider
	Friend Module DataExtensions
		<Extension()>
		Iterator Function ReadAll(ByVal reader As System.Data.IDataReader) As IEnumerable(Of System.Data.IDataRecord)
			While reader.Read()
				Yield reader
			End While
		End Function

		<Extension()>
		Function GetString(ByVal record As System.Data.IDataRecord, ByVal name As String) As String
			Dim value = record.GetValue(record.GetOrdinal(name))
			Return Convert.ToString(value, CultureInfo.InvariantCulture)
		End Function

		<Extension()>
		Function GetInt32(ByVal record As System.Data.IDataRecord, ByVal name As String) As Integer
			Dim value = record.GetValue(record.GetOrdinal(name))
			Return Convert.ToInt32(value, CultureInfo.InvariantCulture)
		End Function

		<Extension()>
		Function GetBoolean(ByVal record As System.Data.IDataRecord, ByVal name As String) As Boolean
			Dim value = record.GetValue(record.GetOrdinal(name))
			Return Convert.ToBoolean(value, CultureInfo.InvariantCulture)
		End Function

		<Extension()>
		Function IsDBNull(ByVal record As System.Data.IDataRecord, ByVal name As String) As Boolean
			Dim value = record.GetValue(record.GetOrdinal(name))
			Return value = DBNull.Value
		End Function
	End Module
End Namespace
